Adding Logging Functionality to Proxy

kb1lqc opened the pull request:
Per #108 I implemented a method of saving data to an SQLite3 database. Largely this will be for testing Telemetry and APRS functionality. I used some methodology discussed with @b00klegger but have minimized the code additions and overall footprint of this code to remain in scope with **only logging data**. We will use separate efforts to actually test the code. This code is needed now to help flesh out how we are automating testing.

Essentially this code is extremely straightforward. I'd suggest stepping through the commits from beginning to end to see the thought process.

When enabled in `proxy.ini`, logging is performed to `log.db` in the proxy folder.

# Database Format
The database is setup in a simplistic manner with a `KeyID` that is unique and auto-increments. Then the Faraday `Port` is saved along with actual `BASE64` data. Lastly, every row is given a computer generated `EPOCH` timestamp indicating when the data was inserted into the database.

# Configuration
A simple boolean value in proxy.ini was added to control logging.

# Logging in Action
As shown below, logging is easily seen with DB Browser for SQLite
![proxylog](https://cloud.githubusercontent.com/assets/331540/24694469/df32d2aa-1996-11e7-85dd-7b9dca7c735e.gif)

kb1lqc committed:
Added initial log boolean to uartworker

Added a simple variable to the uartworker function/thread. I manually set
this in a quick debug statement that will eventually be read from a file.

kb1lqc committed:
Added database section to proxy.sample.ini

Added database section to INI file for sql code to use.

kb1lqc committed:
Added db.sql file for database creation

Correctly sets-up a database for Proxy application logging. This includes
the autoincrementing keyid, Faraday port and BASE64 data, and finally an
EPOCH stamp.

kb1lqc committed:
Added SQLite Database Initialization

Moved over a modified initdb() function from Telemetry application
tailored for Proxy use. The main function was also updated to call this
initialization and sqlite3 was imported.

kb1lqc committed:
Added SQLinsert function and Implemented Logging

Added a modified insert function from Telemetry to log proxy data to
sqlite3. I then implemented this function in uartworker to start saving
data to the database successfully. The insert function required some
simple arranging of dictionary to list for proper insertion order.

kb1lqc committed:
Added correct Logic and INI sample file

Fixed logic to correctly create qeue and detect boolean settings
controlling logging of data to SQLite3. Also added a sample INI file for
proxy which includes logging control properly configured to not log by
default.

kb1lqc committed:
Py.test Updates

I ran py.test and fixed an PEP8/Flake8 errors present in the code.

kb1lqc committed:
Updated Comments

Removed commented out code that was changed with this logging effort. I
also fixed several spelling errors around the code.

kb1lqc committed:
Updated gitignore to ignore database

Ignoring database for storing logged Proxy data.

kb1lqc committed:
Updated start documentation for proxy SQLite3

Added INI file configuration items for SQLite3 database logging code.
Recent commits changed this file and therefore it should be updated.

reillyeon commented on Proxy/proxy.py:
```diff
@@ -71,10 +72,20 @@ def uart_worker(modem, getDicts, units):
                             getDicts[unit][port] = deque([], maxlen=100)
                             getDicts[unit][port].append(item)
 
+                            # Check for Proxy logging and save to SQL if true
```
If we log `item` in either case why not move this to after the try/except block.

kb1lqc replied:
Good point. Fixed

reillyeon commented on Proxy/proxy.py:
```diff
@@ -351,10 +362,106 @@ def callsign2COM():
+
+def sqlInsert(data):
+    """
+    Takes in a data tuple and inserts int into the Proxy SQLite table
```
int -> it

reillyeon commented on Proxy/proxy.py:
```diff
@@ -351,10 +362,106 @@ def callsign2COM():
+
+        # Connect to database, create SQL query, execute query, and close database
+        try:
+            conn = sqlite3.connect(db)
```
`conn` was already opened in the try/except block above.

kb1lqc replied:
Removed. Good catch!

reillyeon commented on Proxy/proxy.py:
```diff
@@ -351,10 +362,106 @@ def callsign2COM():
+
+        except sqlite3.Error as e:
+            logger.error("Sqlite3.Error: " + str(e))
+            conn.rollback()
```
Using `conn` as a context manager will automatically call `rollback()` so it is not necessary here.

kb1lqc replied:
Good catch! [Per the docs](https://docs.python.org/2/library/sqlite3.html#using-the-connection-as-a-context-manager). Removed

kb1lqc committed:
Moved item sql insertion after try/except

@reillyeon pointed out correctly that if we save to the database in both
try and except cases we could just move the saving code to after the
try/except statement. Tested with hardware and works fine.

kb1lqc committed:
Removed unecessary comment

The code is obvious and the comment was lacking any additional clarity and
actually made it less clear referring to "new" buffers. They are qeues.

kb1lqc committed:
Fixed sqlinsert() function docstring

Spelling error in docstring.

kb1lqc committed:
Removed unecessary rollback()

Context manager for SQLite3 automatically performs a rollback if the
execution encounters and exception. Therefore this is unecessary as
@reillyeon points out.

kb1lqd reviewed (approved):
Looks good.
